pkg/test/e2e/konnectivity/konnectivity_test.go

```diff
@@ -0,0 +1,368 @@
+	var err error
+	if userconfig == "" {
+		userconfig, cleanup, err = createUsercluster(t)
+		defer cleanup()
+		if err != nil {
+			t.Fatalf("failed to setup usercluster: %s", spew.Sdump(err))
+		}
+	}
+
+	config, err := clientcmd.BuildConfigFromFlags("", userconfig)
+	if err != nil {
+		t.Fatalf("failed to build config: %s", err)
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		t.Fatalf("failed to create client: %s", err)
+	}
+
+	metricsClient, err := metrics.NewForConfig(config)
+	if err != nil {
+		t.Fatalf("failed to create metrics client: %s", err)
+	}
+
+	//TODO: ensure apiserver sidecar has konnectivity-proxy containers
+
+	ctx := context.Background()
+
+	t.Logf("waiting for nodes to come up")
+	{
+		err := wait.Poll(30*time.Second, 10*time.Minute, func() (bool, error) {
+			t.Logf("checking node readiness...")
+			nodes, err := kubeClient.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+			if err != nil {
+				t.Logf("failed to get nodes list: %s", err)
+				return false, nil
+			}
+			if len(nodes.Items) != 1 {
+				t.Logf(fmt.Sprintf("node count: %d", len(nodes.Items)))
+				return false, nil
+			}
+
+			// assume ready if node has external ip
+			for _, addr := range nodes.Items[0].Status.Addresses {
+				if addr.Type == corev1.NodeExternalIP {
+					return true, nil
+				}
+			}
+			t.Logf("no nodes with external ip")
+			return false, nil
+		})
+		if err != nil {
+			t.Fatalf("nodes never became ready: %v", err)
+		}
+	}
+
+	t.Logf("waiting for pods to get ready")
+	{
+		err := wait.Poll(30*time.Second, 10*time.Minute, func() (bool, error) {
+			t.Logf("checking pod readiness...")
+			pods, err := getPods(ctx, kubeClient, "konnectivity-agent")
+			if err != nil {
+				t.Logf("failed to get pod list: %s", err)
+				return false, nil
+			}
+
+			for _, pod := range pods {
+				if pod.Status.Phase != corev1.PodRunning {
+					t.Logf("not all pods are ready yet...")
+					return false, nil
+				}
+			}
+
+			return true, nil
+		})
+		if err != nil {
+			t.Fatalf("nodes never became ready: %v", err)
+		}
+	}
+
+	t.Log("check if konnectivity-agents are deployed")
+	{
+		pods, err := getPods(ctx, kubeClient, "konnectivity-agent")
+		if err != nil {
+			t.Errorf("failed to get konnectivity-agent pods: %s", err)
+		}
+
+		if len(pods) != 2 {
+			t.Errorf("expected 2 konnectivity-agent pods got: %d", len(pods))
+		}
+	}
+
+	t.Log("check if metric server can talk to kubelet")
+	{
+		pods, err := getPods(ctx, kubeClient, "metrics-server")
+		if err != nil {
+			t.Errorf("failed to get metrics-server pods: %s", err)
+		}
+
+		if len(pods) != 2 {
+			t.Errorf("expected 2 metrics-server pods got: %d", len(pods))
+		}
+
+		for _, pod := range pods {
+			lines := strings.TrimSpace(getPodLogs(ctx, kubeClient, pod))
+			if n := len(strings.Split(lines, "\n")); n != tailLines {
+				t.Fatalf("expected 5 lines got: %d", n)
+			}
+			if !strings.Contains(lines, "Scrape finished") { // TODO: fragile. fix it.
```

Comment:
If this is fragile, maybe we don't have to check for this - as Konnectivity is not involved in metrics-server -> kubelet communication (it is just cluster-local). Checking that metrics-server pods are running may be enough.

Reply:
Ah yes. Removing....